### PR TITLE
Split pytorch autologging into pytorch (for TensorBoard only) and pytorch_lightning (models, ...)

### PR DESCRIPTION
**Pull request by temporaer**
## Related Issues/PRs

#5699

## What changes are proposed in this pull request?

The PR adds basic autolog() functionality for vanilla pytorch by patching pytorch's integrated tensorboard's FileWriter, much like the existing `mlflow.tensorflow._log_event`.

It introduces and registers a new hook for vanilla pytorch, `mlflow.pytorch.autolog()`.
Since this clashes with the existing function of the same name, the PR splits `mlflow.pytorch` into `mlflow.pytorch` and `mlflow.pytorch_lightning`. *This is a breaking change*.

## How is this patch tested?

A test case is in `tests/pytorch/test_pytorch_autolog.py`.

## Does this PR change the documentation?

Docs probably need to be adapted, but I'd first like to be sure that the PR is on solid ground before getting into docs.

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

Add `mlflow.pytorch.autolog_vanilla()` which captures metrics written to `torch.utils.tensorboard.SummaryWriter`.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/pytorch/test_pytorch_lightening_autolog.py [hunk 1]
```diff
@@ -0,0 +1,392 @@
+from packaging.version import Version
```
Comment (temporaer) [inline tests/pytorch/test_pytorch_lightening_autolog.py:1]: note, this is the former test_pytorch_autolog.py, unchanged.

File: mlflow/pytorch/__init__.py [hunk 2]
```diff
@@ -1034,3 +1036,134 @@ def print_auto_logged_info(r):
...
+    <https://pytorch.org/docs/stable/tensorboard.html#torch.utils.tensorboard.writer.SummaryWriter>`_.
+
+    :param log_every_n_epoch: If specified, logs metrics once every `n` epochs. By default, metrics
+                       are logged after every epoch.
+    :param log_every_n_step: If specified, logs batch metrics once every `n` global step.
+                       By default, metrics are not logged for steps. Note that setting this to 1 can
+                       cause performance issues and is not recommended.
+    :param log_models: If ``True``, trained models are logged as MLflow model artifacts.
+                       If ``False``, trained models are not logged.
+    :param disable: If ``True``, disables the PyTorch Lightning autologging integration.
```
Comment (BenWilson2) [inline mlflow/pytorch/__init__.py:1067]:
```suggestion
    :param disable: If ``True``, disables the PyTorch autologging integration.
```
```suggestion
    :param disable: If ``True``, disables the PyTorch autologging integration.
```

Comment (BenWilson2) [inline mlflow/pytorch/__init__.py:1067]:
```suggestion
    :param disable: If ``True``, disables the PyTorch autologging integration.
```
```suggestion
    :param disable: If ``True``, disables the PyTorch autologging integration.
```

File: mlflow/pytorch/__init__.py [hunk 3]
```diff
@@ -1034,3 +1036,134 @@ def print_auto_logged_info(r):
...
+                    If ``False``, enables the PyTorch Lightning autologging integration.
```
Comment (BenWilson2) [inline mlflow/pytorch/__init__.py:1068]:
```suggestion
                    If ``False``, enables the PyTorch autologging integration.
```
```suggestion
                    If ``False``, enables the PyTorch Lightning autologging integration.
```

Comment (BenWilson2) [inline mlflow/pytorch/__init__.py:1068]:
```suggestion
                    If ``False``, enables the PyTorch autologging integration.
```
```suggestion
                    If ``False``, enables the PyTorch Lightning autologging integration.
```

Comment (BenWilson2) [inline mlflow/pytorch/__init__.py:1068]:
```suggestion
                    If ``False``, enables the PyTorch autologging integration.
```
```suggestion
                    If ``False``, enables the PyTorch autologging integration.
```

Comment (BenWilson2) [inline mlflow/pytorch/__init__.py:1068]:
```suggestion
                    If ``False``, enables the PyTorch autologging integration.
```
```suggestion
                    If ``False``, enables the PyTorch autologging integration.
```

File: mlflow/pytorch/__init__.py [hunk 4]
```diff
@@ -1034,3 +1036,134 @@ def print_auto_logged_info(r):
...
+    :param exclusive: If ``True``, autologged content is not logged to user-created fluent runs.
+                      If ``False``, autologged content is logged to the active fluent run,
+                      which may be user-created.
+    :param disable_for_unsupported_versions: If ``True``, disable autologging for versions of
+                      pytorch and pytorch-lightning that have not been tested against this version
```
Comment (BenWilson2) [inline mlflow/pytorch/__init__.py:1073]:
```suggestion
                      pytorch that have not been tested against this version
```
```suggestion
                      pytorch that have not been tested against this version
```

Comment (BenWilson2) [inline mlflow/pytorch/__init__.py:1073]:
```suggestion
                      pytorch that have not been tested against this version
```
```suggestion
                      pytorch that have not been tested against this version
```

File: mlflow/pytorch/__init__.py [hunk 5]
```diff
@@ -1034,3 +1036,134 @@ def print_auto_logged_info(r):
...
+                      of the MLflow client or are incompatible.
+    :param silent: If ``True``, suppress all event logs and warnings from MLflow during PyTorch
```
Comment (BenWilson2) [inline mlflow/pytorch/__init__.py:1075]: Remove lightning references in docstring

Comment (BenWilson2) [inline mlflow/pytorch/__init__.py:1075]: Remove lightning references in docstring

File: mlflow/pytorch/__init__.py [hunk 6]
```diff
@@ -1034,3 +1036,134 @@ def print_auto_logged_info(r):
...
+                                  The registered model is created if it does not already exist.
+
+    .. code-block:: python
+        :caption: Example
+
+        import tempfile
+
+        # Auto log all MLflow entities
+        import mlflow
+        mlflow.pytorch.autolog()
```
Comment (BenWilson2) [inline mlflow/pytorch/__init__.py:1089]: would this be `mlflow.pytorch.autolog_vanilla()` for this example now?

Comment (BenWilson2) [inline mlflow/pytorch/__init__.py:1089]: would this be `mlflow.pytorch.autolog_vanilla()` for this example now?

File: mlflow/utils/autologging_utils/versioning.py [hunk 7]
```diff
@@ -19,7 +19,8 @@
     "tensorflow": ("tensorflow", "tensorflow"),
     "xgboost": ("xgboost", "xgboost"),
     "sklearn": ("sklearn", "sklearn"),
-    "pytorch": ("pytorch_lightning", "pytorch-lightning"),
+    "pytorch_lightening": ("pytorch_lightning", "pytorch-lightning"),
```
Comment (BenWilson2) [inline mlflow/utils/autologging_utils/versioning.py:22]:
```suggestion
    "pytorch_lightning": ("pytorch_lightning", "pytorch-lightning"),
```

Comment (BenWilson2) [inline mlflow/utils/autologging_utils/versioning.py:22]:
```suggestion
    "pytorch_lightning": ("pytorch_lightning", "pytorch-lightning"),
```

File: tests/pytorch/test_pytorch_autolog.py [hunk 8]
```diff
@@ -1,392 +1,28 @@
...
-    assert "betas" in data.params
-
-
-def test_pytorch_autolog_logs_expected_data(pytorch_model):
-    _, run = pytorch_model
-    data = run.data
+    with mlflow.start_run():
+        for i in range(NUM_EPOCHS):
+            writer.add_scalar('loss', 42.0, global_step=i)
+        writer.close()
```
Comment (BenWilson2) [inline tests/pytorch/test_pytorch_autolog.py:21]: Could you add similar test coverage that `pytorch_lightning` has to verify that metrics and params are logged properly with this autologging 'flavor'?

Comment (BenWilson2) [inline tests/pytorch/test_pytorch_autolog.py:21]: Could you add similar test coverage that `pytorch_lightning` has to verify that metrics and params are logged properly with this autologging 'flavor'?

Comment (temporaer) [reply tests/pytorch/test_pytorch_autolog.py:22]: i added a bit more detail to the test.

Comment (temporaer) [reply tests/pytorch/test_pytorch_autolog.py:22]: logging params is a good point, i haven't looked into it.

**Comment by BenWilson2** [comment]
> @BenWilson2 , would you rather go with `autolog_vanilla` or `autolog_tensorboard`?

I think `autolog_tensorboard` is more clear, referencing what setup it supports.

**Comment by BenWilson2** [comment]
Hi @temporaer a gentle ping on this PR

**Comment by temporaer** [comment]
@BenWilson2 , renamed to `mlflow.pytorch.autolog_tensborboard()`.

File: tests/pytorch/test_pytorch_autolog.py (hunk 8 above)
Comment (BenWilson2) [reply tests/pytorch/test_pytorch_autolog.py:25]: 100% agree. Thank you for the additions here!

**Comment by temporaer** [comment]
@BenWilson2 , any more comments on this PR? Btw, is there a way to iterate faster?

**Comment by temporaer** [comment]
friendly ping, @BenWilson2

**Comment by BenWilson2** [comment]
@temporaer terribly sorry on the delay (I've been on paternity leave for the last 8 weeks). I'll take a look at it either today or Tuesday!

**Comment by BenWilson2** [comment]
@temporaer looks like there's some issues with the key types in the flavor test suite for the renaming involved in your PR. Could you address these?

**Comment by temporaer** [comment]
@BenWilson2  no worries, I guess congratulations are due then? 🎉 :-)

**Comment by temporaer** [comment]
@BenWilson2 any suggestions on how to deal with the inconsistency the tests identified? The new interface is called autolog_tensorboard(), not autolog(). Should we split up lightning + "plain" pytorch tensorboard completely?

**Comment by BenWilson2** [comment]
I think they're both going to need the `autolog()` implementation and this breaking change should be handled by a merge to `branch-2.0` rather than to `master`. Can you pull the branch-2.0 changes into your branch and reopen a PR that's merging to `branch-2.0` so we can get this change into the MLflow 2.0 release?

**Comment by BenWilson2** [comment]
@temporaer gentle nudge :)

**Comment by temporaer** [comment]
@BenWilson2  looks like CI has an issue here, DCO says there are >40 commits but there are only 2 i'd like to push?

**Comment by temporaer** [comment]
@BenWilson2 , it looks like the `_add_to_queue` function in mlflow.tensorboard that i was relying on is gone now in the 2.0 branch. While searching for the new mechanism, i discovered that there seems to be no mechanism outside of test code that writes to the `_metric_queue`. Is this intentional?

**Comment by temporaer** [comment]
@BenWilson2 there's one failing test left, but I'm having trouble setting up things locally to reproduce: creating the pyenv fails since pip can't find the (non-existent) `transformers==4.20.0.dev0` version. I can't figure out where the requiremeent is coming from. Could you have a look please, either at the test or the repro issue? thank you!

**Comment by akshaya-a** [comment]
@BenWilson2  @dbczumar  sorry I can't think of a popular tool that is less intuitive than git so I'm not really sure what happened with this PR - there is some activity that looks like there were commits here to get it in line with 2.0 but I don't think I see the final code on master though this PR is closed. Is that intended?

**Comment by dbczumar** [comment]
Hi @temporaer  - apologies for the conflicts here with `branch-2.0`. Would you be able to re-file it against `master`? @BenWilson2 can you help here?

**Comment by temporaer** [comment]
I targeted master originally, Ben asked me to rebase on V2. Can you confirm the ask?

**Comment by dbczumar** [comment]
Ah, the 2.0 branch was since merged into `master` in preparation for the MLflow 2.0 release. It may be easier to check out the pytorch-specific changes from a previous commit on your branch (prior to the 2.0 merge). Again, so sorry for the hassle here.

**Comment by BenWilson2** [comment]
The branch close for this was automated (sorry for this! We had to recut that branch for release). @temporaer let me know how I can help here.

**Comment by temporaer** [comment]
@BenWilson2 as this is a breaking change, if it won't go into V2 it'd have to wait for V3, is this correct? I'd rather not spend time on it now in this case.

**Comment by dbczumar** [comment]
> @BenWilson2 as this is a breaking change, if it won't go into V2 it'd have to wait for V3, is this correct? I'd rather not spend time on it now in this case.

We can definitely work with you to structure this in a way that isn't breaking. If we begin with the breaking PR, we can provide adjustment suggestions to ensure that this will roll out in MLflow 2.1.

**Comment by temporaer** [comment]
I see. For starters, if V2 could simply rename pytorch to pytorch lightning, it would be much easier to integrate the current pr later. So if you're in a hurry to release v2, we can split it up into two parts?

**Pull request closed without merging**